### ISSUE TITLE
Problem occured  when training with Custom dataset and test with the dataset

### ISSUE THREAD

**Issue by DGUnggi** [ultralytics/ultralytics#11717]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

`!pip install PyYAML

import yaml

data = {'train' : '/workspace/ExDark-12/train/images/',
        'val' : '/workspace/ExDark-12/valid/images/',
        'test' : '/workspace/ExDark-12/test/images/',
        'names' : ['bicycle','boat','bottle','bus','cat','cup','motorbike','person','table','car','chair','dog'],
        'nc' : 12}

with open('/workspace/ExDark-12/data.yaml', 'w') as f:
  yaml.dump(data, f)

with open('/workspace/ExDark-12/data.yaml', 'r') as f:
  exdark_yaml = yaml.safe_load(f)
  display(exdark_yaml)
model = YOLO('yolov8m.pt')
model.train(data = '/workspace/ExDark-12/data.yaml', epochs = 50, batch = 32, imgsz = 640, name = "Best", exist_ok = True )
`
![image](https://github.com/ultralytics/ultralytics/assets/129149629/730e6e57-1add-42b1-a367-e35ff90a11ba)
![image](https://github.com/ultralytics/ultralytics/assets/129149629/a3a4a6ca-b321-4967-8f94-bf7f266a4615)
![image](https://github.com/ultralytics/ultralytics/assets/129149629/98698412-1601-419a-beaf-5aa22783b7a4)
As you can see above the picture, i want to test the model and evaluate the ability of the model, but test mode did not work.
What is the reason and how to solve the problem? 
I really tried to solve the issue, but i cannot did this...

**Comment by glenn-jocher** [comment]
Hello! It looks like you've done a thorough job setting up your dataset and initiating training with YOLOv8. 🚀

For testing your model and evaluating its performance, the process slightly differs from training. You'd want to use the `val` mode to evaluate the model on your test dataset. Ensure that your `data.yaml` includes the correct path under `test`. Here's an example snippet to help you begin the evaluation process:

```python
from ultralytics import YOLO

# Load your trained model
model = YOLO('/workspace/ExDark-12/weights/best.pt')  # Adjust the path to your model weights

# Evaluate the model
results = model.val(data='/workspace/ExDark-12/data.yaml')
print("Evaluation metrics:", results)
```

This will evaluate the model using the settings and dataset specified in your YAML file. Ensure the test images are accessible at the path specified under the `test` key in your YAML file.

If the `val` command does not work, consider verifying the path to the test images and ensuring that the dataset is correctly formatted. If issues persist, please share any error messages or additional details to help diagnose the problem more effectively.

Good luck with your testing! 🌟

**Comment by DGUnggi** [comment]
@glenn-jocher Thank you !
I have a question. What is difference between `results = model.val(data='/workspace/ExDark-12/data.yaml')` and `results = model.val(data='/workspace/ExDark-12/data.yaml', split = 'test')`
I think this is the reason why the result is different.

**Comment by glenn-jocher** [comment]
Hello! Great question! When you run `model.val(data='/workspace/ExDark-12/data.yaml')`, it will default to using the validation dataset specified in your `data.yaml` under the `val` key. On the other hand, when you specify `split='test'` as in `model.val(data='/workspace/ExDark-12/data.yaml', split='test')`, it tells the model to evaluate using the test dataset specified under the `test` key in your `data.yaml`. This switch is indeed why you see different results, as it's testing on different subsets of your data. Hope this clarifies things! 😊

**Comment by DGUnggi** [comment]
@glenn-jocher Thank you for your detailed explanation!
I have another problem. When i eval the original YOLOv8 with dataset whose number of classes is 7 and name according to each index is below. 
'names' : ['bicycle','bus','car','motorcycle','person','rider', 'truck']
However, YOLOv8 model class index is below.
  "class": {
    "0": "person",
    "1": "bicycle",
    "2": "car",
    "3": "motorcycle",
    "4": "airplane",
    "5": "bus",
    "6": "train",
    "7": "truck",
    "8": "boat",
    "9": "traffic light",
    "10": "fire hydrant",
    "11": "stop sign",
    "12": "parking meter",
    "13": "bench",
    "14": "bird",
    "15": "cat",
    "16": "dog",
    "17": "horse",
    "18": "sheep",
    "19": "cow",
    "20": "elephant",
    "21": "bear",
    "22": "zebra",
    "23": "giraffe",
    "24": "backpack",
    "25": "umbrella",
    "26": "handbag",
    "27": "tie",
    "28": "suitcase",
    "29": "frisbee",
    "30": "skis",
    "31": "snowboard",
    "32": "sports ball",
    "33": "kite",
    "34": "baseball bat",
    "35": "baseball glove",
    "36": "skateboard",
    "37": "surfboard",
    "38": "tennis racket",
    "39": "bottle",
    "40": "wine glass",
    "41": "cup",
    "42": "fork",
    "43": "knife",
    "44": "spoon",
    "45": "bowl",
    "46": "banana",
    "47": "apple",
    "48": "sandwich",
    "49": "orange",
    "50": "brocolli",
    "51": "carrot",
    "52": "hot dog",
    "53": "pizza",
    "54": "donut",
    "55": "cake",
    "56": "chair",
    "57": "couch",
    "58": "potted plant",
    "59": "bed",
    "60": "dining table",
    "61": "toilet",
    "62": "tv",
    "63": "laptop",
    "64": "mouse",
    "65": "remote",
    "66": "keyboard",
    "67": "cell phone",
    "68": "microwave",
    "69": "oven",
    "70": "toaster",
    "71": "sink",
    "72": "refrigerator",
    "73": "book",
    "74": "clock",
    "75": "vase",
    "76": "scissors",
    "77": "teddy bear",
    "78": "hair drier",
    "79": "toothbrush"
  }
I think that not matching class index causes the result wrong. this is the result.
![image](https://github.com/ultralytics/ultralytics/assets/129149629/0465ef7d-29b8-4c32-aae8-4b3a753c15e8)
How can i solve this..?

**Comment by DGUnggi** [comment]
@glenn-jocher 
In this case, should i make a new accuracy and precision function?

**Comment by glenn-jocher** [comment]
Hello! Indeed, if the class indices in your model do not directly match those in your target evaluation dataset, creating a custom function to map and align these indices before calculating accuracy and precision would be beneficial. Here’s a simple way to do that adjustment:

```python
# Assuming 'preds' are your model's predictions and 'index_map' is defined as in the previous example
adjusted_preds = [index_map[pred] if pred in index_map else -1 for pred in preds]

# Now, use 'adjusted_preds' to calculate accuracy and precision
```

Just ensure you handle any unmapped or misaligned classes appropriately to maintain the integrity of your evaluation metrics. If you have any more questions or need further assistance, feel free to ask! 😊

**Comment by DGUnggi** [comment]
@glenn-jocher 
I search the YOLOv8 Ultralytics to postprocess the result. 
In this way, what is the attribute(ex) result object) should i use to make a custom function? 
`result = model(input)` In this code, there are many attributes and methods like masks, cpu(), to(), etc..
But i cannot find the prediction result(class index for each object)

**Comment by DGUnggi** [comment]
@glenn-jocher 
In addition, can i calculate the accuracy and precision? If so, how can i calculate? In Ultralytics, it didn't gives me result for each one object, one image, but provides total result.

**Comment by glenn-jocher** [comment]
Hey there! Great question! 😊 If your class indices don't directly match those expected by YOLOv8 due to custom modifications, you might indeed need to adjust your predictions as discussed earlier. After this step, you can then calculate accuracy and precision manually to ensure they correctly reflect the performance against your specific classes.

Here's a quick example of how you might calculate precision manually using adjusted predictions:

```python
from sklearn.metrics import precision_score

# true_labels should be your ground truth labels
# adjusted_preds are your predictions after index mapping
precision = precision_score(true_labels, adjusted_preds, average='macro')
print(f"Precision: {precision}")
```

Adjusting the predictions ensures that your evaluations are accurate and reflective of your specific dataset configuration. Let me know if you need further assistance!

**Comment by DGUnggi** [comment]
@glenn-jocher Thank you for your supporting! I completed this work successfully!

**Comment by glenn-jocher** [comment]
That's fantastic to hear! 🎉 If you have any more questions or need further assistance as you continue your work, feel free to reach out. Happy coding! 😊
